ntoskrnl/io/pnpmgr/pnpnotify.c

```diff
@@ -19,210 +19,274 @@
 typedef struct _PNP_NOTIFY_ENTRY
 {
     LIST_ENTRY PnpNotifyList;
-    IO_NOTIFICATION_EVENT_CATEGORY EventCategory;
     PVOID Context;
-    UNICODE_STRING Guid;
-    PFILE_OBJECT FileObject;
     PDRIVER_OBJECT DriverObject;
     PDRIVER_NOTIFICATION_CALLBACK_ROUTINE PnpNotificationProc;
+    union
+    {
+        GUID Guid; // for EventCategoryDeviceInterfaceChange
+        struct // for EventCategoryTargetDeviceChange
+        {
+            PFILE_OBJECT FileObject;
+        };
+    };
+    IO_NOTIFICATION_EVENT_CATEGORY EventCategory;
+    volatile LONG Deleted;
 } PNP_NOTIFY_ENTRY, *PPNP_NOTIFY_ENTRY;
 
-KGUARDED_MUTEX PnpNotifyListLock;
-LIST_ENTRY PnpNotifyListHead;
+KGUARDED_MUTEX IopTargetDeviceNotifyLock;
+KGUARDED_MUTEX IopDeviceClassNotifyLock;
+KGUARDED_MUTEX IopHwProfileNotifyLock;
+
+LIST_ENTRY PiNotifyHwProfileListHead;
+LIST_ENTRY PiNotifyDeviceInterfaceListHead;
 
 /* FUNCTIONS *****************************************************************/
 
+CODE_SEG("INIT")
 VOID
-IopNotifyPlugPlayNotification(
-    IN PDEVICE_OBJECT DeviceObject,
-    IN IO_NOTIFICATION_EVENT_CATEGORY EventCategory,
-    IN LPCGUID Event,
-    IN PVOID EventCategoryData1,
-    IN PVOID EventCategoryData2)
+PiInitializeNotifications(VOID)
 {
-    PPNP_NOTIFY_ENTRY ChangeEntry;
-    PLIST_ENTRY ListEntry;
-    PVOID NotificationStructure;
-    BOOLEAN CallCurrentEntry;
-    UNICODE_STRING GuidString;
-    NTSTATUS Status;
-    PDEVICE_OBJECT EntryDeviceObject = NULL;
+    KeInitializeGuardedMutex(&IopTargetDeviceNotifyLock);
+    KeInitializeGuardedMutex(&IopDeviceClassNotifyLock);
+    KeInitializeGuardedMutex(&IopHwProfileNotifyLock);
+    InitializeListHead(&PiNotifyHwProfileListHead);
+    InitializeListHead(&PiNotifyDeviceInterfaceListHead);
+}
 
-    ASSERT(DeviceObject);
+CODE_SEG("PAGE")
+VOID
+PiFreeTargetDeviceNotifyList(
+    _In_ PDEVICE_NODE DeviceNode)
+{
+    PAGED_CODE();
 
-    KeAcquireGuardedMutex(&PnpNotifyListLock);
-    if (IsListEmpty(&PnpNotifyListHead))
+    for (PLIST_ENTRY entry = DeviceNode->TargetDeviceNotify.Flink;
+         entry != &DeviceNode->TargetDeviceNotify;
+         entry = entry->Flink)
     {
-        KeReleaseGuardedMutex(&PnpNotifyListLock);
+        PPNP_NOTIFY_ENTRY nEntry = CONTAINING_RECORD(entry, PNP_NOTIFY_ENTRY, PnpNotifyList);
+        ASSERT(nEntry->Deleted);
+
+        RemoveEntryList(&nEntry->PnpNotifyList);
+        ObDereferenceObject(nEntry->DriverObject);
+        ExFreePoolWithTag(nEntry, TAG_PNP_NOTIFY);
+    }
+}
+
+static
+CODE_SEG("PAGE")
+VOID
+PiProcessSingleNotification(
+    _In_ PPNP_NOTIFY_ENTRY Entry,
+    _In_ IO_NOTIFICATION_EVENT_CATEGORY EventCategory,
+    _In_ PVOID NotificationStructure,
+    _In_ PKGUARDED_MUTEX Lock)
+{
+    PDEVICE_INTERFACE_CHANGE_NOTIFICATION notifyStruct = NotificationStructure;
+
+    PAGED_CODE();
+
+    // actually remove "deleted" entries
+    if (InterlockedAnd(&Entry->Deleted, 1))
```

Comment:
What's the point of ```InterlockedAnd``` here?

Reply:
I used it as an interlocked load operation